### PR TITLE
[win32] Refactor toolbar dpi callback

### PR DESCRIPTION
**Pull request by akoch-yatta**
This PR mainly is targeted on refactoring the existing callback for ToolBar and ToolItem after a DPI change in the win32 implementation. 
To achieve that a method to scale an ImageList was added. Scaled variants of an image list are saved and reused similar to the other Resources.
The refactoring uses the scalable ImageLists to reuse the existing toolbar items and only resets the ImageList with a correctly scaled one. Due to the limitations of the win32 implementation all items must be removed and re-added via the win32-API to make shrinking of toolbars possible. Additionally, the commit removes the now unused callback in ToolItem.

All calls to the ImageList handle are replaced to use the added method.

### CODE REVIEW

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java [hunk 1]
```diff
@@ -23,6 +25,9 @@ public class ImageList {
 	int style, refCount;
 	Image [] images;
 	private final int zoom;
+	int width, height;
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java:28]: Can these values be private?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java [hunk 2]
```diff
@@ -445,4 +457,22 @@ public int size () {
 	return result;
 }
 
+public static long win32_getHandle(ImageList imageList, int zoom) {
+	if(!imageList.zoomToHandle.containsKey(zoom)) {
+		int scaledWidth = DPIUtil.scaleUp(DPIUtil.scaleDown( imageList.width,imageList.zoom), zoom);
+		int scaledHeight = DPIUtil.scaleUp(DPIUtil.scaleDown( imageList.height,imageList.zoom), zoom);
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java:463]: whitespace seems to be broken here, e.g.
`DPIUtil.scaleDown( imageList.width,imageList.zoom)` -> `DPIUtil.scaleDown(imageList.width, imageList.zoom)`

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java [hunk 3]
```diff
@@ -445,4 +457,22 @@ public int size () {
...
+		long handle = OS.ImageList_Create (scaledWidth, scaledHeight, OS.ILC_MASK | OS.ILC_COLOR32, 16, 16);
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java:464]: Shouldnt't this use the flags of the original list?
```suggestion
		long handle = OS.ImageList_Create (scaledWidth, scaledHeight, imageList.flags, 16, 16);
```

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java [hunk 4]
```diff
@@ -445,4 +457,22 @@ public int size () {
 	return result;
 }
 
+public static long win32_getHandle(ImageList imageList, int zoom) {
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java:460]: Shouldn't the `getHandle()` method be removed in favor of the `win32_getHandle()` method? It should even be possible to replace the existing `getHandle()` method with one accepting a zoom parameter, as this class is internal and its interface may be changed.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/ToolBar.java [hunk 5]
```diff
@@ -1742,75 +1745,59 @@ private static void handleDPIChange(Widget widget, int newZoom, float scalingFac
...
+	var seperatorWidth = new int[toolItems.length];
+	int itemCount = toolItems.length;
+
+	// Remove and re-add all button the let Windows resize the tool bar
+    ToolItem[] items = new ToolItem[itemCount];
+    TBBUTTON[] buttondata = new TBBUTTON[itemCount];
+    for (int i = itemCount - 1; i >= 0; i--) {
+    	ToolItem item = toolItems[i];
+    	if ((item.style & SWT.SEPARATOR) != 0 && item.getControl() != null) {
+    		// Take note of widths of separators with control, so they can be resized
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/ToolBar.java:1758]: There seem to be some inconsistencies in the formatting of this file.

**Review by HeikoKlare — COMMENTED** [review]
I have not tested the proposal yet, but I have taken a first look at the code. So far, I only have minor comments.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java (hunk 1 above)
Comment (akoch-yatta) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java]: Yes, I made all added variables private

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java (hunk 2 above)
Comment (akoch-yatta) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java]: Yes, I adapted that in the latest commit

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java (hunk 3 above)
Comment (akoch-yatta) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java]: Good point. We store the flags in the image list now and use that for the creation of the scaled variant

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java (hunk 4 above)
Comment (akoch-yatta) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java]: Yes, definitely. I moved the new logic into getHandle and adapted all calls accordingly

**Review by HeikoKlare — APPROVED** [review]
The change looks sound to me. I've also tested it (in combination with https://github.com/akoch-yatta/eclipse.platform.ui/commit/384e390ac8e9e4e4a912695880aaef864de07897 to properly see effects on the main toolbar) and found it working fine.

@fedejeanne maybe you could also have a look since parts of the change slightly affect the behavior also when using existing HiDPI support, so it would be good to have an additional review.

**Review by fedejeanne — APPROVED** [review]
LGTM, just 1 nit

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java [hunk 6]
```diff
@@ -23,16 +25,23 @@ public class ImageList {
 	int style, refCount;
 	Image [] images;
 	private final int zoom;
+	private int width, height;
+	private int flags;
```
[the remark this answers is not in the document]
Comment (akoch-yatta) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/ImageList.java]: I adapted that in [ee0b8d](https://github.com/eclipse-platform/eclipse.platform.swt/commit/ee0b8d7f887dc269d53e69efa13857c4a9508031)

**Pull request merged**